zjit/src/hir.rs

```diff
@@ -3199,6 +3278,75 @@ mod validation_tests {
+        let val = function.push_insn(function.entry_block, Insn::ArrayDup { val: dangling, state: InsnId(0usize) });
+        let fun_string = format!("{:?}", function);
+        assert_matches_err(function.validate_definite_assignment(), ValidationError::OperandNotDefined(fun_string, entry, val, dangling));
+    }
+
+    #[test]
+    fn using_non_output_insn() {
+        let mut function = Function::new(std::ptr::null());
+        let entry = function.entry_block;
+        let const_ = function.push_insn(function.entry_block, Insn::Const{val: Const::CBool(true)});
+        // Ret is a non-output instruction.
+        let ret = function.push_insn(function.entry_block, Insn::Return { val: const_ });
+        let val = function.push_insn(function.entry_block, Insn::ArrayDup { val: ret, state: InsnId(0usize) });
+        let fun_string = format!("{:?}", function);
+        assert_matches_err(function.validate_definite_assignment(), ValidationError::OperandNotDefined(fun_string, entry, val, ret));
+    }
+
+    #[test]
+    fn not_dominated_by_diamond() {
+        // This tests that one branch is missing a definition which fails.
+        let mut function = Function::new(std::ptr::null());
+        let entry = function.entry_block;
+        let side = function.new_block();
+        let exit = function.new_block();
+        let param1 = function.push_insn(side, Insn::Param { idx: 1 });
+        let v0 = function.push_insn(entry, Insn::Const { val: Const::Value(VALUE::fixnum_from_usize(3)) });
+        function.push_insn(side, Insn::Jump(BranchEdge { target: exit, args: vec![] }));
+        let val1 = function.push_insn(entry, Insn::Const { val: Const::CBool(false) });
+        function.push_insn(entry, Insn::IfFalse { val: val1, target: BranchEdge { target: side, args: vec![v0] } });
+        function.push_insn(entry, Insn::Jump(BranchEdge { target: exit, args: vec![v0] }));
+        let param0 = function.push_insn(exit, Insn::Param { idx: 0 });
+        let val2 = function.push_insn(exit, Insn::ArrayDup { val: param1, state: param0 });
+        crate::cruby::with_rubyvm(|| {
+            function.infer_types();
+        });
+        let fun_string = format!("{:?}", function);
+        assert_matches_err(function.validate_definite_assignment(), ValidationError::OperandNotDefined(fun_string, exit, val2, param1));
+    }
+
+    #[test]
+    fn dominated_by_diamond() {
+        // This tests that both branches with a definition succeeds.
+        let mut function = Function::new(std::ptr::null());
+        let entry = function.entry_block;
+        let side = function.new_block();
+        let exit = function.new_block();
+        let param1 = function.push_insn(side, Insn::Param { idx: 1 });
+        let v0 = function.push_insn(side, Insn::Const { val: Const::Value(VALUE::fixnum_from_usize(3)) });
+        function.push_insn_id(entry, v0);
+        function.push_insn(side, Insn::Jump(BranchEdge { target: exit, args: vec![param1] }));
```

Comment:
This is testing that things get passed by parameter but not testing that in A -> B, A -> C, B -> D, C -> D situation that D can use an InsnId from A. If that's intentional due to backend constraints right now, great --- please document it and add a TODO to allow "reaching backward" to dominating blocks. But for that we wouldn't need dataflow analysis at all; it would all be block-local. So instead I think you should test something that doesn't use block params